tracks/rust/exercises/bob/mentoring.md

```diff
@@ -0,0 +1,140 @@
+### Concepts
+
+- [bool](https://doc.rust-lang.org/std/primitive.bool.html)
+- [char](https://doc.rust-lang.org/std/char/index.html)
+- [match statement](https://doc.rust-lang.org/1.2.0/book/match.html)
+- [variables](https://doc.rust-lang.org/book/ch03-01-variables-and-mutability.html)
+- [closures](https://doc.rust-lang.org/book/ch13-01-closures.html)
+- [&str](https://doc.rust-lang.org/std/primitive.str.html)
+- [Iterators](https://doc.rust-lang.org/std/iter/trait.Iterator.html)
+- [Iterator Methods](https://doc.rust-lang.org/std/iter/trait.Iterator.html#provided-methods)
+- [constant](https://doc.rust-lang.org/std/keyword.const.html)
+
+### Reasonable Solutions
+
+A reasonable solution should:
+
+- Use constants
+- Use variables
+- Use `match` statements
+- Check if the input string is empty and return the "Sure" constant
+before doing any additional processing
+- Use simple methods from the standard library
+- Not use the regex crate
+- Not use separate functions
+
+#### Examples
+
+```rust
+const SURE: &str = "Sure.";
+const WHOA: &str = "Whoa, chill out!";
+const CALM: &str = "Calm down, I know what I'm doing!";
+const FINE: &str = "Fine. Be that way!";
+const WHATEVER: &str = "Whatever.";
+
+#[allow(clippy::match_bool)]
+pub fn reply(message: &str) -> &str {
+    let trimmed: &str = message.trim();
+
+    match trimmed.is_empty() {
+        true => FINE,
+        false => {
+            let is_question: bool = trimmed.ends_with("?");
+
+            let contains_alphabetic_characters: bool = trimmed.chars().any(|character| character.is_alphabetic());
+            let is_uppercase: bool = trimmed == trimmed.to_uppercase();
+            let is_yelled: bool = contains_alphabetic_characters && is_uppercase;
+
+            match (is_question, is_yelled) {
+                (true, false) => SURE,
+                (false, true) => WHOA,
+                (true, true) => CALM,
+                (false, false) => WHATEVER,
+            }
+        }
+    }
+}
+```
+
+### Common Suggestions
+
+If they don't use a constant:
+```
+We can improve this by setting our output strings to be [constants](https://doc.rust-lang.org/std/keyword.const.html)
+```
+
+If they don't create a trimmed variable or use something other than `.trim()`:
+```
+Have a look at the `.trim()` method.
+```
+
+If they check if the string is empty in any other way than `.is_empty()`:
+```
+Well done you for seeing that we need to check if the string is empty with `== ""`. There's actually a method, `.is_empty()`, that we can use instead, which is much cleaner.
+```
+
+If they don't use `match`:
+```
+Check out using `match` instead of `if`. For some things, like destructuring, it's required. It's an idiom specific to Rust, so I always default to it unless the logic is comparing many different things and a match in that case makes it messy.
+```
+
+If they use an `if` on their `.is_empty()` check:
```

Comment:
I disagree quite a bit here. Doing an early return, and not having the rest in an arm is a win for readability here imho (subjective of course), and lessens the dreaded rightwards drift that sometimes plagues rust code. I'd say both variants are totally fine and you should remove this suggestion.

Reply:
I use `match` as much as possible as I find it more readable. So, I suggest it to my students and if they prefer the `if` statement, that's fine, but most of them didn't think about it as an option because of the pressures of taking on a new language and having clippy tell them that it is "wrong", and are grateful for the suggestion :)

Quoting from [clippy](https://rust-lang.github.io/rust-clippy/master/index.html#match_bool), "It makes the code less readable.". Being that it is purely a stylistic choice, this gives the student the power to make their own decisions about how they feel their code looks best.

Reply:
It should therefore be a talking point (tips, hints, brain farts) and not a suggestion (usually things that make a solution disprovable).